forms.rst

```diff
@@ -113,119 +138,236 @@ from inside a controller::
+defining them in controller actions. Besides, forms defined in classes can be
+reused in multiple actions and services.
 
-Finally, you added a submit button with a custom label for submitting the form to
-the server.
+Form classes are :ref:`form types <form-types>` that implement
+:class:`Symfony\\Component\\Form\\FormTypeInterface`. However, it's better to
+extend from :class:`Symfony\\Component\\Form\\AbstractType`, which already
+implements the interface and provides some utilities::
 
-Symfony comes with many built-in types that will be discussed shortly
-(see :ref:`forms-type-reference`).
+    // src/Form/Type/TaskType.php
+    namespace App\Form;
 
-.. index::
-  single: Forms; Basic template rendering
+    use Symfony\Component\Form\AbstractType;
+    use Symfony\Component\Form\Extension\Core\Type\DateType;
+    use Symfony\Component\Form\Extension\Core\Type\SubmitType;
+    use Symfony\Component\Form\Extension\Core\Type\TextType;
+    use Symfony\Component\Form\FormBuilderInterface;
 
-Rendering the Form
-~~~~~~~~~~~~~~~~~~
+    class TaskType extends AbstractType
+    {
+        public function buildForm(FormBuilderInterface $builder, array $options)
+        {
+            $builder
+                ->add('task', TextType::class)
+                ->add('dueDate', DateType::class)
+                ->add('save', SubmitType::class)
+            ;
+        }
+    }
+
+The form class contains all the directions needed to create the task form. In
+controllers extending from the :ref:`AbstractController <the-base-controller-class-services>`,
+use the ``createForm()`` helper (otherwise, use the ``create()`` method of the
+``form.factory`` service)::
+
+    // src/Controller/TaskController.php
+    use App\Form\Type\TaskType;
+
+    public function new()
+    {
+        // creates a task object and initializes some data for this example
+        $task = new Task();
+        $task->setTask('Write a blog post');
+        $task->setDueDate(new \DateTime('tomorrow'));
+
+        $form = $this->createForm(TaskType::class, $task);
+
+        // ...
+    }
+
+.. _form-data-class:
+
+Every form needs to know the name of the class that holds the underlying data
+(e.g. ``App\Entity\Task``). Usually, this is just guessed based off of the
+object passed to the second argument to ``createForm()`` (i.e. ``$task``).
+Later, when you begin :doc:`embedding forms </form/embedded>`, this will no
+longer be sufficient.
+
+So, while not always necessary, it's generally a good idea to explicitly specify
+the ``data_class`` option by adding the following to your form type class::
+
+        // src/Form/Type/TaskType.php
+        use App\Entity\Task;
+        use Symfony\Component\OptionsResolver\OptionsResolver;
+
+        // ...
+
+        public function configureOptions(OptionsResolver $resolver)
+        {
+            $resolver->setDefaults([
+                'data_class' => Task::class,
+            ]);
+        }
```

Comment:
This example has too much identation and I always prefer to put the `class ... { }` around methods. It seems easier to understand